charger/fbdect.go

```diff
@@ -0,0 +1,353 @@
+	updated                       time.Time
+}
+
+func init() {
+	registry.Add("fbdect", NewFbDectFromConfig)
+}
+
+// NewFbDectFromConfig creates a fbdect charger from generic config
+func NewFbDectFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		URI      string
+		AIN      string
+		User     string
+		Password string
+		SID      string
+		Cache    time.Duration
+	}{
+		URI:      "",
+		AIN:      "",
+		User:     "",
+		Password: "",
+		SID:      "",
+		Cache:    0,
+	}
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	if cc.URI == "" || cc.AIN == "" {
+		return nil, errors.New("fbdect config: must have uri and ain of AVM FritzDECT switch")
+	}
+
+	return NewFbDect(cc.URI, cc.AIN, cc.User, cc.Password, cc.SID, cc.Cache)
+}
+
+// NewFbDect creates FbDect charger
+func NewFbDect(uri, ain, user, password, sid string, cache time.Duration) (*FbDect, error) {
+	c := &FbDect{
+		Helper:   request.NewHelper(util.NewLogger("fbdect")),
+		uri:      strings.TrimRight(uri, "/"),
+		ain:      strings.TrimSpace(ain),
+		user:     strings.TrimSpace(user),
+		password: strings.TrimSpace(password),
+		sid:      strings.TrimSpace(sid),
+	}
+
+	return c, nil
+}
+
+func (c *FbDect) fbResponse(function string) string {
+
+	loginURL := fmt.Sprintf("%s/webservices/homeautoswitch.lua", c.uri)
+	parameters := make(map[string]string)
+	// Refresh Fritzbox session id
+	if time.Since(c.updated).Minutes() >= 10 {
+		c.getFbSessionID()
+		// Update session timestamp
+		c.updated = time.Now()
+	}
+
+	parameters["sid"] = c.sid
+	if c.ain != "" {
+		parameters["ain"] = c.ain
+	}
+	parameters["switchcmd"] = function
+
+	return strings.TrimSpace(string(sendFbRequest(loginURL, parameters)))
+}
+
+func (c *FbDect) apiStatus() (status fbdectStatusResponse, err error) {
+	status.Name = c.fbResponse("getswitchname")
+	status.Present, err = strconv.ParseInt(c.fbResponse("getswitchpresent"), 10, 64)
+	if err != nil {
+		return status, err
+	}
+	status.State, err = strconv.ParseInt(c.fbResponse("getswitchstate"), 10, 32)
+	if err != nil {
+		return status, err
+	}
+	status.Temperature, err = strconv.ParseFloat(c.fbResponse("gettemperature"), 64)
+	if err != nil {
+		return status, err
+	}
+	status.Temperature = status.Temperature / 10
+	if err != nil {
+		return status, err
+	}
+	status.Power, err = strconv.ParseFloat(c.fbResponse("getswitchpower"), 64)
+	if err != nil {
+		return status, err
+	}
+	status.Power = status.Power / 1000 // mW ==> W
+	status.Energy, err = strconv.ParseFloat(c.fbResponse("getswitchenergy"), 64)
+	if err != nil {
+		return status, err
+	}
+	status.Energy = status.Energy / 1000 // Wh ==> kWh
+
+	return status, err
+}
+
+// apiUpdate invokes fbdect api
+func (c *FbDect) apiUpdate(function string) (status fbdectStatusResponse, err error) {
+	if function == "SetSwitchOn" {
+		status.State, err = strconv.ParseInt(strings.TrimSpace(c.fbResponse("setswitchon")), 10, 32)
```

Comment:
mach das trimspace in die fbresponse, anstatt 100x im Code dupliziert?

Reply:
Hatte ich schon drin, nur vergessen den AUfruf anzupassen.